From 2e3f382a38463f28e0807e5c5fb042a61db629da Mon Sep 17 00:00:00 2001
From: Yu Chien Peter Lin <peterlin@andestech.com>
Date: Mon, 8 May 2023 16:26:33 +0800
Subject: [PATCH 149/172] lib: utils: cache: Add andes_l2c driver

Implement andes_l2c_enable/wbinval_all/get_addr function.

Reformed from the following patches on ast-v5_3_0-branch
- (775311967516) lib: utils/cache: Add andes_l2c driver
- (4237cc84d94b) lib: utils/cache: Fix riscv32 address data type (#68)
- (de729b20aa1e) andes: andes_sbi: Fix wbinval_all warning when L2C does  not exist
- (e653f16a9ff5) lib: utils/cache: Add "get_addr" method for cache framework

Signed-off-by: Hui Min Mina Chou <minachou@andestech.com>
---
 lib/utils/cache/Kconfig                  |   8 +
 lib/utils/cache/fdt_cache_andes_l2c.c    | 182 +++++++++++++++++++++++
 lib/utils/cache/objects.mk               |   3 +
 platform/generic/andes/ae350.c           |   2 +
 platform/generic/configs/andes_defconfig |   2 +
 platform/generic/include/andes/andes.h   |   2 +
 6 files changed, 199 insertions(+)
 create mode 100644 lib/utils/cache/fdt_cache_andes_l2c.c

diff --git a/lib/utils/cache/Kconfig b/lib/utils/cache/Kconfig
index a5d0540cc4fe..6898a2df0289 100644
--- a/lib/utils/cache/Kconfig
+++ b/lib/utils/cache/Kconfig
@@ -8,6 +8,14 @@ config FDT_CACHE
 	select CACHE
 	default n
 
+if FDT_CACHE
+
+config FDT_CACHE_ANDES_L2C
+	bool "Andes L2 cache controller"
+	default n
+
+endif
+
 config CACHE
 	bool "Cache controller support"
 	default n
diff --git a/lib/utils/cache/fdt_cache_andes_l2c.c b/lib/utils/cache/fdt_cache_andes_l2c.c
new file mode 100644
index 000000000000..a226ec3dcf0d
--- /dev/null
+++ b/lib/utils/cache/fdt_cache_andes_l2c.c
@@ -0,0 +1,182 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2023 Andes Technology Corporation
+ *
+ * Authors:
+ *   Yu Chien Peter Lin <peterlin@andestech.com>
+ */
+
+#include <sbi/riscv_asm.h>
+#include <sbi/riscv_io.h>
+#include <sbi/sbi_bitops.h>
+#include <sbi/sbi_console.h>
+#include <sbi/sbi_error.h>
+#include <sbi/sbi_types.h>
+#include <sbi_utils/cache/cache.h>
+#include <sbi_utils/cache/fdt_cache.h>
+#include <sbi_utils/fdt/fdt_helper.h>
+#include <andes/andes.h>
+
+/* clang-format off */
+
+/* L2 cache registers */
+#define L2C_REG_CFG_OFFSET		0
+#define L2C_REG_CTL_OFFSET		0x8
+#define L2C_HPM_C0_CTL_OFFSET		0x10
+#define L2C_HPM_C1_CTL_OFFSET		0x18
+#define L2C_HPM_C2_CTL_OFFSET		0x20
+#define L2C_HPM_C3_CTL_OFFSET		0x28
+#define L2C_REG_C0_CMD_OFFSET		0x40
+#define L2C_REG_C0_ACC_OFFSET		0x48
+#define L2C_REG_C1_CMD_OFFSET		0x50
+#define L2C_REG_C1_ACC_OFFSET		0x58
+#define L2C_REG_C2_CMD_OFFSET		0x60
+#define L2C_REG_C2_ACC_OFFSET		0x68
+#define L2C_REG_C3_CMD_OFFSET		0x70
+#define L2C_REG_C3_ACC_OFFSET		0x78
+#define L2C_REG_C0_STATUS_OFFSET	0x80
+#define L2C_REG_C0_HPM_OFFSET		0x200
+
+/* Per-hart offsets */
+#define CCTL_CMD_REG(hart)	(L2C_REG_C0_CMD_OFFSET + (hart) * (l2c.cmd_stride))
+#define CCTL_STATUS_REG(hart)	(L2C_REG_C0_STATUS_OFFSET + (hart) * (l2c.status_stride))
+#define CCTL_L2_STATUS_C0_MASK	0xF
+#define CCTL_L2_STATUS_CN_MASK(hart)	\
+	(CCTL_L2_STATUS_C0_MASK << ((hart) * l2c.status_bit_offset))
+
+/* Configuration register encoding */
+#define MEM_MAP_OFF	20
+#define MEM_MAP_MSK	BIT(MEM_MAP_OFF)
+
+/* Control register encoding */
+#define L2_ENABLE	0x1
+
+/* L2 CCTL status */
+#define CCTL_L2_STATUS_IDLE		0
+#define CCTL_L2_STATUS_PROCESS		1
+#define CCTL_L2_STATUS_ILLEGAL		2
+
+/* L2 cache operation */
+#define CCTL_L2_WBINVAL_ALL		0x12
+
+/* clang-format on */
+
+struct l2c_data {
+	unsigned long addr;
+	/* L2C Gen1/Gen2 quirks */
+	unsigned int cmd_stride;
+	unsigned int status_stride;
+	unsigned int status_bit_offset;
+};
+
+static struct l2c_data l2c;
+
+static int andes_l2c_enable(void)
+{
+	u32 ctrl;
+
+	ctrl = readl((void *)(l2c.addr + L2C_REG_CTL_OFFSET));
+	if (!(ctrl & L2_ENABLE))
+		writel(ctrl | L2_ENABLE,
+		       (void *)(l2c.addr + L2C_REG_CTL_OFFSET));
+
+	/* Check if L2C is enabled */
+	return (readl((void *)(l2c.addr + L2C_REG_CTL_OFFSET)) & L2_ENABLE)
+	       ? 0
+	       : SBI_EINVAL;
+}
+
+static inline uint32_t l2c_get_cctl_status(u32 hartid)
+{
+	return readl((void *)(l2c.addr + CCTL_STATUS_REG(hartid)));
+}
+
+static int andes_l2c_wbinval_all(void)
+{
+	u32 hartid;
+
+	if (!(readl((void *)(l2c.addr + L2C_REG_CTL_OFFSET)) & L2_ENABLE))
+		return SBI_EINVAL; /* L2C is not enabled */
+
+	/*
+	 * All harts share the same L2 cache. Selecting the current hart
+	 * to do the WBINVAL.
+	 */
+	hartid = current_hartid();
+	writel(CCTL_L2_WBINVAL_ALL, (void *)(l2c.addr + CCTL_CMD_REG(hartid)));
+	/* Wait for L2C to complete the command */
+	while ((l2c_get_cctl_status(hartid) & CCTL_L2_STATUS_CN_MASK(hartid))
+	       != CCTL_L2_STATUS_IDLE)
+		continue;
+
+	return 0;
+}
+
+static int andes_l2c_get_addr(unsigned long *addr)
+{
+	*addr = l2c.addr;
+	return 0;
+}
+
+static struct cache andes_l2c = {
+	.enable	     = andes_l2c_enable,
+	.wbinval_all = andes_l2c_wbinval_all,
+	.get_addr    = andes_l2c_get_addr,
+};
+
+static int andes_l2c_init(void *fdt, int nodeoff, const struct fdt_match *match)
+{
+	int rc;
+	uint64_t addr;
+	ulong mmsc_cfg;
+
+	rc = fdt_get_node_addr_size(fdt, nodeoff, 0, &addr, NULL);
+	if (rc)
+		return rc;
+
+	/* On 45 and newer CPU series, we can also check with mmsc_cfg.L2C bit */
+	if (!(is_andes(25) || is_andes(27))) {
+#if __riscv_xlen == 64
+		mmsc_cfg = csr_read(CSR_MMSC_CFG);
+		if (!EXTRACT_FIELD(mmsc_cfg, CSR_MMSC_CFG_L2C_MASK)) {
+#else
+		mmsc_cfg = csr_read(CSR_MMSC_CFG2);
+		if (!EXTRACT_FIELD(mmsc_cfg, CSR_MMSC_CFG2_L2C_MASK)) {
+#endif
+			sbi_printf(
+				"%s: WARNING: The Andes platform doesn't support L2C, the l2-cache@%lx node should not be valid.\n",
+				__func__, (unsigned long)addr);
+			return SBI_ENODEV;
+		}
+	}
+
+	l2c.addr = (unsigned long)addr;
+	if (readl((void *)(l2c.addr + L2C_REG_CFG_OFFSET)) & MEM_MAP_MSK) {
+		/* v1 memory map (Gen2) */
+		l2c.cmd_stride	      = 0x1000;
+		l2c.status_stride     = 0x1000;
+		l2c.status_bit_offset = 0x0;
+	} else {
+		/* v0 memory map (Gen1) */
+		l2c.cmd_stride	      = 0x10;
+		l2c.status_stride     = 0x0;
+		l2c.status_bit_offset = 0x4;
+	}
+
+	rc = cache_add(&andes_l2c);
+	if (rc)
+		return rc;
+
+	return 0;
+}
+
+static const struct fdt_match andes_l2c_match[] = {
+	{ .compatible = "cache" },
+	{},
+};
+
+struct fdt_cache fdt_cache_andes_l2c = {
+	.match_table = andes_l2c_match,
+	.init	     = andes_l2c_init,
+};
diff --git a/lib/utils/cache/objects.mk b/lib/utils/cache/objects.mk
index 258acaf632a8..a3f240be9fb6 100644
--- a/lib/utils/cache/objects.mk
+++ b/lib/utils/cache/objects.mk
@@ -11,3 +11,6 @@ libsbiutils-objs-$(CONFIG_CACHE) += cache/cache.o
 
 libsbiutils-objs-$(CONFIG_FDT_CACHE) += cache/fdt_cache.o
 libsbiutils-objs-$(CONFIG_FDT_CACHE) += cache/fdt_cache_drivers.o
+
+carray-fdt_cache_drivers-$(CONFIG_FDT_CACHE_ANDES_L2C) += fdt_cache_andes_l2c
+libsbiutils-objs-$(CONFIG_FDT_CACHE_ANDES_L2C) += cache/fdt_cache_andes_l2c.o
diff --git a/platform/generic/andes/ae350.c b/platform/generic/andes/ae350.c
index 87e38781283e..e59527b4b474 100644
--- a/platform/generic/andes/ae350.c
+++ b/platform/generic/andes/ae350.c
@@ -9,6 +9,7 @@
 
 #include <platform_override.h>
 #include <andes/andes_pmu.h>
+#include <sbi_utils/cache/fdt_cache.h>
 #include <sbi_utils/fdt/fdt_helper.h>
 #include <sbi_utils/fdt/fdt_fixup.h>
 #include <sbi_utils/sys/atcsmu.h>
@@ -113,6 +114,7 @@ static int ae350_final_init(bool cold_boot, const struct fdt_match *match)
 	pma_init();
 	ae350_hsm_device_init();
 	trigger_init();
+	fdt_cache_init();
 
 	return 0;
 }
diff --git a/platform/generic/configs/andes_defconfig b/platform/generic/configs/andes_defconfig
index 30df912859c9..f1155f16183c 100644
--- a/platform/generic/configs/andes_defconfig
+++ b/platform/generic/configs/andes_defconfig
@@ -9,3 +9,5 @@ CONFIG_FDT_SERIAL=y
 CONFIG_FDT_SERIAL_UART8250=y
 CONFIG_FDT_TIMER=y
 CONFIG_FDT_TIMER_PLMT=y
+CONFIG_FDT_CACHE=y
+CONFIG_FDT_CACHE_ANDES_L2C=y
diff --git a/platform/generic/include/andes/andes.h b/platform/generic/include/andes/andes.h
index 823b27c1edca..1b825c4d2cf1 100644
--- a/platform/generic/include/andes/andes.h
+++ b/platform/generic/include/andes/andes.h
@@ -60,6 +60,8 @@
 #define CSR_MMSC_CFG2		0xfc3
 #define CSR_MVEC_CFG		0xfc7
 #define CSR_MMSC_CFG_PFT_MASK	(1 << 4)
+#define CSR_MMSC_CFG_L2C_MASK	(1UL << 46)
+#define CSR_MMSC_CFG2_L2C_MASK	(1UL << 14)
 
 /* Counter Related Registers */
 #define CSR_MCOUNTERWEN		0x7ce
-- 
2.34.1

